### PR TITLE
Dashboard privilege API and collaborators UI

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

Follow-up to https://github.com/PostHog/posthog/pull/8394 adding an API endpoint to control dashboard privileges and `ShareModal` UI to manage collaborators. **Don't merge**, I'll handle it in time to avoid merge conflicts resulting from squash merging in the wrong order.

~~The UI currently looks like this.~~ See https://github.com/PostHog/posthog/pull/8424#issuecomment-1030292228 for UI.

## How did you test this code?

Django API tests.

### CODE REVIEW

**Comment by Twixes** [comment]
Alright, so this PR basically adds this collaborators management flow (addressing to-dos "API to add and remove dashboard collaborators (with solid tests)" and "UI to add and remove dashboard collaborators" of https://github.com/PostHog/posthog/pull/8394):

**Review by clarkus — CHANGES_REQUESTED** [review]
Looks good to me. The remove actions shouldn't be inline, though. I don't think removing users is going to be a super frequent action, so it's placed as a distinct action in the drop down. 

Screen Shot 2022-02-04 at 11 56 17 AM

**Comment by Twixes** [comment]
Hmm, for now "can view" is exactly equivalent with being removed though, which is pretty confusing. The "can view/can edit" dropdown will make sense once there's the option of "Only those invited can view" on the dashboard, but the initial iteration of the feature focuses on editing restrictions. @clarkus

**Comment by clarkus** [comment]
> Hmm, for now "can view" is exactly equivalent with being removed though, which is pretty confusing. The "can view/can edit" dropdown will make sense once there's the option of "Only those invited can view" on the dashboard, but the initial iteration of the feature focuses on editing restrictions. @clarkus

Got it! That makes sense. We've been using trash can icons for inline remove actions elsewhere, but I can see how that might be seen as deleting a user in this context. Let's go with what you have until we get the full set of permissions added.

**Review by clarkus — APPROVED** [review]

**Comment by pauldambra** [comment]
Fake Paul?! I'm scandalised 🤣

File: posthog/api/plugin.py [hunk 1]
```diff
@@ -175,7 +175,7 @@ def get_queryset(self):
         return queryset.none()
 
     def filter_queryset_by_parents_lookups(self, queryset):
-        parents_query_dict = self.get_parents_query_dict()
+        parents_query_dict = self.parents_query_dict
```
Comment (pauldambra) [inline posthog/api/plugin.py:178]: does this allocation exist for a reason or could it be inlined?

**Review by pauldambra — APPROVED** [review]

File: posthog/api/plugin.py (hunk 1 above)
Comment (Twixes) [reply posthog/api/plugin.py:178]: Definitely can be inlined now!

**Review by mariusandra — COMMENTED** [review]
I managed to get this state:



No matter if I shared the dashboard or not, Fake Paul could see it.

Permissions were as such:

image

Going to continue reviewing, but did I do something obviously wrong here?

**Review by mariusandra — APPROVED** [review]
LGTM 👍

**Comment by Twixes** [comment]
For context, we chatted on Slack and it was a matter of:
1. The actual restrictions only being reflected in a follow-up PR (https://github.com/PostHog/posthog/pull/8462).
2. The restriction setting button not conveying intuitively that the choice is specifically around **editing** restrictions, and not viewing. Details and a proposed alternative design for this [in Figma](https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=7378%3A35806).

**Pull request merged**